File: backend/app/models/multiplayer.py
from typing import Any, Literal

from pydantic import BaseModel, Field


RoomStatus = Literal["waiting", "started", "finished", "closed"]
ParticipantRole = Literal["host", "participant"]


class ParticipantState(BaseModel):
    id: str
    name: str
    role: ParticipantRole
    is_host: bool = False
    joined_at: float
    completed: bool = False
    score: int | None = None
    accuracy: int | None = None
    average_response_time: int | None = None


class RoomState(BaseModel):
    code: str
    status: RoomStatus
    host_id: str
    participants: list[ParticipantState]
    participant_count: int
    created_at: float
    started_at: float | None = None
    quiz: dict[str, Any] | None = None
    settings: dict[str, Any] | None = None
    leaderboard: list[ParticipantState] = Field(default_factory=list)
    all_finished: bool = False

File: frontend/types/multiplayer.ts
export type ParticipantRole = "host" | "participant";
export type RoomStatus = "waiting" | "started" | "finished" | "closed";

export type ParticipantState = {
  id: string;
  name: string;
  role: ParticipantRole;
  is_host: boolean;
  joined_at: number;
  completed: boolean;
  score: number | null;
  accuracy: number | null;
  average_response_time: number | null;
};

export type RoomState = {
  code: string;
  status: RoomStatus;
  host_id: string;
  participants: ParticipantState[];
  participant_count: number;
  created_at: number;
  started_at: number | null;
  quiz: import("@/types/quiz").GeneratedQuiz | null;
  settings: import("@/types/quiz").QuizSettings | null;
  leaderboard: ParticipantState[];
  all_finished: boolean;
};

export type RoomAck = {
  ok: boolean;
  message?: string;
  participant_id?: string;
  room?: RoomState;
};
